mono/utils/mono-os-mutex.h

```diff
@@ -366,29 +379,26 @@ mono_os_cond_init (mono_cond_t *cond)
 static inline void
 mono_os_cond_destroy (mono_cond_t *cond)
 {
-	/* Beauty of win32 API: do not destroy it */
-}
-
-static inline void
-mono_os_cond_wait (mono_cond_t *cond, mono_mutex_t *mutex)
-{
-	BOOL res;
-
-	res = SleepConditionVariableCS (cond, mutex, INFINITE);
-	if (G_UNLIKELY (res == 0))
-		g_error ("%s: SleepConditionVariableCS failed with error %d", __func__, GetLastError ());
+	// There is no way to destroy a Win32 condition variable.
 }
 
 static inline int
 mono_os_cond_timedwait (mono_cond_t *cond, mono_mutex_t *mutex, guint32 timeout_ms)
 {
-	BOOL res;
+	const BOOL res = mutex->recursive ?
+		SleepConditionVariableCS (cond, &mutex->critical_section, timeout_ms) :
+		SleepConditionVariableSRW (cond, &mutex->srwlock, timeout_ms, 0);
 
-	res = SleepConditionVariableCS (cond, mutex, timeout_ms);
-	if (G_UNLIKELY (res == 0 && GetLastError () != ERROR_TIMEOUT))
-		g_error ("%s: SleepConditionVariableCS failed with error %d", __func__, GetLastError ());
+	if (!G_LIKELY (res || (timeout_ms != INFINITE && GetLastError () == ERROR_TIMEOUT)))
```

Comment:
I guess anyone could have passed INFINITE before, and I should not strengthen the assert this way.